core/src/mindustry/mod/Scripts.java

```diff
@@ -116,6 +118,46 @@ public Music loadMusic(String soundName){
         return sound;
     }
 
+    /** Ask the user to select a file to read for a certain purpose like "Please upload a sprite"
+        Extension can be null to select any file. */
+    public void readFile(String purpose, String ext, Cons<String> cons){
+        selectFile(true, purpose, ext, fi -> cons.get(fi.readString()));
+    }
+
+    /** readFile but for a byte[] */
+    public void readBinFile(String purpose, String ext, Cons<byte[]> cons){
+        selectFile(true, purpose, ext, fi -> cons.get(fi.readBytes()));
+    }
+
+    /** Ask the user to write a file.
+        Extension can be null to allow any filename. */
+    public void writeFile(String purpose, String ext, String contents){
+        if(contents == null) contents = "";
+        final String fContents = contents;
+        selectFile(false, purpose, ext, fi -> fi.writeString(fContents));
+    }
+
+    /** writeFile but for a byte[] */
+    public void writeBinFile(String purpose, String ext, byte[] contents){
+        if(contents == null) contents = new byte[0];
+        final byte[] fContents = contents;
+        selectFile(false, purpose, ext, fi -> fi.writeBytes(fContents));
+    }
+
+    private void selectFile(boolean open, String purpose, String ext, Cons<Fi> cons){
+        purpose = purpose.startsWith("@") ? Core.bundle.get(purpose.substring(1)) : purpose;
+        //add purpose and extension at the top
+        String title = Core.bundle.get(open ? "@open": "@save") + " - " + purpose + (ext == null ? "" : " (." + ext + ")");
+        Vars.platform.showFileChooser(open, title, ext, fi -> {
```

Comment:
You can't have a null extension supplied to the file chooser. On desktop platforms, it'll probably display nothing, and on other platforms it may just crash.

Reply:
would `""` function how i want null to?

Reply:
For desktop: No.
For iOS/Android: There are multiple file chooser implementations and it's hard to figure out what will happen unless you test it manually.

Reply:
ah well, everything has extensions set in stone might as well go with it.